File: reference/python/ssklib/parse_sskb.py
"""Parse .sskb binary files."""

import math
import struct

import yaml

from .error import SSKError
from .parse_ssk import _StrictLoader, _check_properties, _check_root

_MAGIC = b'SSKB'
_SHAPES = {0: 'circle', 1: 'ngon'}
_MODES  = {0: 'add', 1: 'subtract', 2: 'intersect'}

# field_mask bit positions
_B_POINTS     = 0
_B_ROTATION   = 1
_B_SIZE       = 2
_B_SHAPE      = 3
_B_SIDES      = 4
_B_MODE       = 5
_B_AFFECTS    = 6
_B_PROPERTIES = 7

_MIN_PIECE_BYTES = 11
_MIN_POINT_BYTES = 18
_ROOT_PROPERTY_LENGTH_BYTES = 4

class _Reader:
    __slots__ = ('_buf', '_pos')

    def __init__(self, data: bytes):
        self._buf = data
        self._pos = 0

    def _need(self, n: int):
        if self._pos + n > len(self._buf):
            raise SSKError("truncated sskb input")

    def remaining(self) -> int:
        return len(self._buf) - self._pos

    def require_count(self, count: int, min_item_size: int, ctx: str, *, reserved_tail: int = 0):
        if self.remaining() < reserved_tail:
            raise SSKError("truncated sskb input")
        available = self.remaining() - reserved_tail
        if count > available // min_item_size:
            raise SSKError(f"{ctx}: count {count} exceeds remaining input")

    def u8(self) -> int:
        self._need(1)
        v = self._buf[self._pos]
        self._pos += 1
        return v

    def u16(self) -> int:
        self._need(2)
        v = struct.unpack_from('<H', self._buf, self._pos)[0]
        self._pos += 2
        return v

    def u32(self) -> int:
        self._need(4)
        v = struct.unpack_from('<I', self._buf, self._pos)[0]
        self._pos += 4
        return v

    def f32(self) -> float:
        self._need(4)
        v = struct.unpack_from('<f', self._buf, self._pos)[0]
        self._pos += 4
        if math.isnan(v) or math.isinf(v):
            raise SSKError("non-finite f32 value in sskb")
        return v

    def raw(self, n: int) -> bytes:
        self._need(n)
        v = self._buf[self._pos:self._pos + n]
        self._pos += n
        return v

    def vec3(self) -> dict:
        return {'x': self.f32(), 'y': self.f32(), 'z': self.f32()}

    def vec2(self) -> dict:
        return {'x': self.f32(), 'y': self.f32()}

    def done(self) -> bool:
        return self._pos >= len(self._buf)


def _bit(mask: int, pos: int) -> bool:
    return (mask >> pos) & 1 != 0


def _read_prop_blob(r: _Reader, ctx: str, *, empty_as_none: bool = True):
    n = r.u32()
    if n == 0:
        return None if empty_as_none else {}
    raw = r.raw(n)
    try:
        text = raw.decode('utf-8')
    except UnicodeDecodeError as e:
        raise SSKError(f"property blob not valid UTF-8: {e}")
    try:
        props = yaml.load(text, Loader=_StrictLoader)
    except yaml.YAMLError as e:
        raise SSKError(f"{ctx}: malformed property blob: {e}")
    if not isinstance(props, dict):
        raise SSKError(f"{ctx}: property blob must be a YAML mapping")
    _check_properties(props, ctx)
    return props


def _read_point(r: _Reader) -> dict:
    pt = {}
    v = r.vec3()
    pt['x'], pt['y'], pt['z'] = v['x'], v['y'], v['z']

    if r.u8(): pt['curve_in']       = r.vec3()
    if r.u8(): pt['curve_out']      = r.vec3()
    if r.u8(): pt['size']           = r.vec3()
    if r.u8(): pt['rotation']       = r.vec3()
    if r.u8(): pt['transition_in']  = r.vec2()
    if r.u8(): pt['transition_out'] = r.vec2()
    return pt


def _read_piece(r: _Reader) -> dict:
    piece = {}
    piece['id'] = r.u32()
    has_from = r.u8()

    if has_from:
        piece['from'] = r.u32()
        fm = r.u16()
        if fm & 0xFF00:
            raise SSKError(f"piece {piece['id']}: reserved field_mask bits set")
    else:
        fm = 0xFD  # all fields except rotation are always present

    # points
    if not has_from or _bit(fm, _B_POINTS):
        n = r.u32()
        r.require_count(n, _MIN_POINT_BYTES, f"piece {piece['id']} points")
        piece['points'] = [_read_point(r) for _ in range(n)]

    # rotation
    if not has_from:
        if r.u8():
            piece['rotation'] = r.vec3()
    elif _bit(fm, _B_ROTATION):
        piece['rotation'] = r.vec3()

    # size
    if not has_from or _bit(fm, _B_SIZE):
        piece['size'] = r.vec3()

    # shape
    if not has_from or _bit(fm, _B_SHAPE):
        sv = r.u8()
        if sv not in _SHAPES:
            raise SSKError(f"piece {piece['id']}: invalid shape enum {sv}")
        piece['shape'] = _SHAPES[sv]

    # sides
    if not has_from:
        if r.u8():
            piece['sides'] = r.u32()
    elif _bit(fm, _B_SIDES):
        piece['sides'] = r.u32()

    # mode : always encoded for non-inherited; default add is omitted
    if not has_from or _bit(fm, _B_MODE):
        mv = r.u8()
        if mv not in _MODES:
            raise SSKError(f"piece {piece['id']}: invalid mode enum {mv}")
        name = _MODES[mv]
        if has_from or name != 'add':
            piece['mode'] = name

    # affects
    if not has_from:
        if r.u8():
            n = r.u32()
            r.require_count(n, 4, f"piece {piece['id']} affects")
            piece['affects'] = [r.u32() for _ in range(n)]
    elif _bit(fm, _B_AFFECTS):
        n = r.u32()
        r.require_count(n, 4, f"piece {piece['id']} affects")
        piece['affects'] = [r.u32() for _ in range(n)]

    # properties
    if not has_from or _bit(fm, _B_PROPERTIES):
        props = _read_prop_blob(
            r,
            f"piece {piece['id']} properties",
            empty_as_none=not has_from,
        )
        if props is not None:
            piece['properties'] = props

    return piece

# public API

def parse(data: bytes) -> dict:

    r = _Reader(data)

    magic = r.raw(4)
    if magic != _MAGIC:
        raise SSKError(f"bad sskb magic: expected {_MAGIC!r}, got {magic!r}")

    major, minor = r.u16(), r.u16()
    if major > 1:
        raise SSKError(f"unsupported sskb major version: {major}")

    n = r.u32()
    r.require_count(n, _MIN_PIECE_BYTES, "pieces", reserved_tail=_ROOT_PROPERTY_LENGTH_BYTES)
    pieces = [_read_piece(r) for _ in range(n)]
    root_props = _read_prop_blob(r, "root properties")

    if not r.done():
        raise SSKError("extra trailing bytes in sskb")

    doc = {'pieces': pieces}
    if root_props is not None:
        doc['properties'] = root_props
    _check_root(doc)
    return doc

File: implementation/packages/typescript/src/parseSskb.ts
import { parseSsk, checkProperties, checkRoot } from './parseSsk.js';
import { SSKError } from './error.js';
import type { Piece, PropertyValue, SSKDocument } from './types.js';

const MAGIC = [0x53, 0x53, 0x4b, 0x42];
const SHAPES = new Map([[0, 'circle'], [1, 'ngon']] as const);
const MODES = new Map([[0, 'add'], [1, 'subtract'], [2, 'intersect']] as const);

const B_POINTS = 0;
const B_ROTATION = 1;
const B_SIZE = 2;
const B_SHAPE = 3;
const B_SIDES = 4;
const B_MODE = 5;
const B_AFFECTS = 6;
const B_PROPERTIES = 7;

const MIN_PIECE_BYTES = 11;
const MIN_POINT_BYTES = 18;
const ROOT_PROPERTY_LENGTH_BYTES = 4;

export function parseSskb(data: ArrayBuffer | Uint8Array): SSKDocument {
  const reader = new Reader(data);
  const magic = [...reader.raw(4)];
  if (!MAGIC.every((value, index) => value === magic[index])) throw new SSKError(`bad sskb magic: expected b'SSKB', got ${JSON.stringify(magic)}`);

  const major = reader.u16();
  const minor = reader.u16();
  if (major > 1) throw new SSKError(`unsupported sskb major version: ${major}`);

  const count = reader.u32();
  reader.requireCount(count, MIN_PIECE_BYTES, 'pieces', ROOT_PROPERTY_LENGTH_BYTES);
  const pieces = Array.from({ length: count }, () => readPiece(reader));
  const rootProps = readPropBlob(reader, 'root properties');

  if (!reader.done()) throw new SSKError('extra trailing bytes in sskb');
  const doc: SSKDocument = { pieces };
  if (rootProps !== undefined) doc.properties = rootProps as Record<string, PropertyValue>;
  checkRoot(doc as unknown as Record<string, unknown>);
  void minor;
  return doc;
}

class Reader {
  private readonly bytes: Uint8Array;
  private readonly view: DataView;
  private pos = 0;

  constructor(data: ArrayBuffer | Uint8Array) {
    this.bytes = data instanceof Uint8Array ? data : new Uint8Array(data);
    this.view = new DataView(this.bytes.buffer, this.bytes.byteOffset, this.bytes.byteLength);
  }

  remaining(): number {
    return this.bytes.length - this.pos;
  }

  requireCount(count: number, minItemSize: number, ctx: string, reservedTail = 0): void {
    if (this.remaining() < reservedTail) throw new SSKError('truncated sskb input');
    const available = this.remaining() - reservedTail;
    if (count > Math.floor(available / minItemSize)) throw new SSKError(`${ctx}: count ${count} exceeds remaining input`);
  }

  u8(): number {
    this.need(1);
    return this.bytes[this.pos++];
  }

  u16(): number {
    this.need(2);
    const value = this.view.getUint16(this.pos, true);
    this.pos += 2;
    return value;
  }

  u32(): number {
    this.need(4);
    const value = this.view.getUint32(this.pos, true);
    this.pos += 4;
    return value;
  }

  f32(): number {
    this.need(4);
    const value = this.view.getFloat32(this.pos, true);
    this.pos += 4;
    if (!Number.isFinite(value)) throw new SSKError('non-finite f32 value in sskb');
    return value;
  }

  raw(count: number): Uint8Array {
    this.need(count);
    const value = this.bytes.slice(this.pos, this.pos + count);
    this.pos += count;
    return value;
  }

  vec3() {
    return { x: this.f32(), y: this.f32(), z: this.f32() };
  }

  vec2() {
    return { x: this.f32(), y: this.f32() };
  }

  done(): boolean {
    return this.pos >= this.bytes.length;
  }

  private need(count: number): void {
    if (this.pos + count > this.bytes.length) throw new SSKError('truncated sskb input');
  }
}

function bit(mask: number, pos: number): boolean {
  return ((mask >> pos) & 1) !== 0;
}

function readPropBlob(reader: Reader, ctx: string, emptyAsNone = true): Record<string, PropertyValue> | undefined {
  const count = reader.u32();
  if (count === 0) return emptyAsNone ? undefined : {};
  const raw = reader.raw(count);
  let text: string;
  try {
    text = new TextDecoder('utf-8', { fatal: true }).decode(raw);
  } catch (error) {
    throw new SSKError(`property blob not valid UTF-8: ${(error as Error).message}`);
  }
  let props: unknown;
  try {
    props = parseSsk(`pieces: []\nproperties:\n${indentYaml(text)}`).properties;
  } catch (error) {
    throw new SSKError(`${ctx}: malformed property blob: ${(error as Error).message}`);
  }
  if (!props || typeof props !== 'object' || Array.isArray(props)) throw new SSKError(`${ctx}: property blob must be a YAML mapping`);
  checkProperties(props, ctx);
  return props as Record<string, PropertyValue>;
}

function indentYaml(text: string): string {
  return text.split(/\r?\n/).map((line) => line ? `  ${line}` : line).join('\n');
}

function readPoint(reader: Reader) {
  const point: Record<string, unknown> = reader.vec3();
  if (reader.u8()) point.curve_in = reader.vec3();
  if (reader.u8()) point.curve_out = reader.vec3();
  if (reader.u8()) point.size = reader.vec3();
  if (reader.u8()) point.rotation = reader.vec3();
  if (reader.u8()) point.transition_in = reader.vec2();
  if (reader.u8()) point.transition_out = reader.vec2();
  return point;
}

function readPiece(reader: Reader): Piece {
  const piece: Record<string, unknown> = { id: reader.u32() };
  const hasFrom = reader.u8() !== 0;
  let fieldMask = 0xfd;
  if (hasFrom) {
    piece.from = reader.u32();
    fieldMask = reader.u16();
    if (fieldMask & 0xff00) throw new SSKError(`piece ${piece.id}: reserved field_mask bits set`);
  }

  if (!hasFrom || bit(fieldMask, B_POINTS)) {
    const count = reader.u32();
    reader.requireCount(count, MIN_POINT_BYTES, `piece ${piece.id} points`);
    piece.points = Array.from({ length: count }, () => readPoint(reader));
  }
  if (!hasFrom) {
    if (reader.u8()) piece.rotation = reader.vec3();
  } else if (bit(fieldMask, B_ROTATION)) {
    piece.rotation = reader.vec3();
  }
  if (!hasFrom || bit(fieldMask, B_SIZE)) piece.size = reader.vec3();
  if (!hasFrom || bit(fieldMask, B_SHAPE)) {
    const value = reader.u8();
    const shape = SHAPES.get(value as 0 | 1);
    if (!shape) throw new SSKError(`piece ${piece.id}: invalid shape enum ${value}`);
    piece.shape = shape;
  }
  if (!hasFrom) {
    if (reader.u8()) piece.sides = reader.u32();
  } else if (bit(fieldMask, B_SIDES)) {
    piece.sides = reader.u32();
  }
  if (!hasFrom || bit(fieldMask, B_MODE)) {
    const value = reader.u8();
    const mode = MODES.get(value as 0 | 1 | 2);
    if (!mode) throw new SSKError(`piece ${piece.id}: invalid mode enum ${value}`);
    if (hasFrom || mode !== 'add') piece.mode = mode;
  }
  if (!hasFrom) {
    if (reader.u8()) {
      const count = reader.u32();
      reader.requireCount(count, 4, `piece ${piece.id} affects`);
      piece.affects = Array.from({ length: count }, () => reader.u32());
    }
  } else if (bit(fieldMask, B_AFFECTS)) {
    const count = reader.u32();
    reader.requireCount(count, 4, `piece ${piece.id} affects`);
    piece.affects = Array.from({ length: count }, () => reader.u32());
  }
  if (!hasFrom || bit(fieldMask, B_PROPERTIES)) {
    const props = readPropBlob(reader, `piece ${piece.id} properties`, !hasFrom);
    if (props !== undefined) piece.properties = props;
  }
  return piece as Piece;
}
